src/vsr/replica.zig

```diff
@@ -7250,6 +7251,24 @@ pub fn ReplicaType(
                 return;
             }
 
+            // To avoid falsely contributing to the durability of the current checkpoint, replicas
+            // syncing table blocks should not send prepare_oks for prepares past the current
+            // checkpoint's prepare_max. However, ops past prepare_max + pipeline_prepare_queue_max
+            // can be safely prepare_ok'd. This is because if the primary prepared these ops,
+            // prepare_max + 1 is committed and the checkpoint is durable on a commit quorum!
+            const prepare_max_current_checkpoint =
+                vsr.Checkpoint.prepare_max_for_checkpoint(self.op_checkpoint());
+            if (!self.sync_content_done() and header.op > prepare_max_current_checkpoint.? and
+                (header.op <= prepare_max_current_checkpoint.? +
+                constants.pipeline_prepare_queue_max))
```

Comment:
This is ugly fomatting! Not a problem per se, but usually a sign that there's some out of the box simplification possible. 

The trivial improvement here would be to split the if into the principal and helper condition, and to push `.?` up:

```zig
            if (!self.sync_content_done()) {
                // To avoid falsely contributing to the durability of the current checkpoint,
                // replicas syncing table blocks should not send prepare_oks for prepares past the
                // current checkpoint's prepare_max. However, ops past prepare_max +
                // pipeline_prepare_queue_max can be safely prepare_ok'd. This is because if the
                // primary prepared these ops, prepare_max + 1 is committed and the checkpoint is
                // durable on a commit quorum!
                const prepare_max_current_checkpoint =
                    vsr.Checkpoint.prepare_max_for_checkpoint(self.op_checkpoint()).?;

                if (header.op > prepare_max_current_checkpoint and
                    (header.op <= prepare_max_current_checkpoint + constants.pipeline_prepare_queue_max))
                {
                    log.debug("{}: send_prepare_ok: not sending (sync_content_done={})", .{
                        self.replica,
                        self.sync_content_done(),
                    });
                    return;
                }
            }
```

This is more readable, but actually makes line length problem worse. 

If we look at the offending condition about the pipeline, it becomes clear that something else is wrong here --- the usage of pipeline length constant is somewhat arbitrarty here. Like, if you think about it, it becomes clear why this implies what you think it should imply, but it requires some indirect reasons. I think a more direct way to say this would be 

```zig
                if (header.op > prepare_max_current_checkpoint and
                    (self.commit_max <= prepare_max_current_checkpoint))
```

?

That way it's exactly the condition from op_repair_min:

```
     const prepare_max_for_current_checkpoint =
                    vsr.Checkpoint.prepare_max_for_checkpoint(self.op_checkpoint()).?;
                // After state sync, commit_max might lag behind checkpoint_op.
                maybe(self.commit_max < prepare_max_for_current_checkpoint);
                if (self.commit_max > prepare_max_for_current_checkpoint) {
           
```

Reply:
Agreed, this certainly isn't the most aesthetically pleasing way to do this. I will clean it up, thanks a lot for the suggestions!

I _had_ considered just using `self.commit_max > prepare_max_for_current_checkpoint` here (similar to op_repair_min), but the thought got lost somewhere in a sea of ideas. 🤦🏼

Reply:
https://github.com/tigerbeetle/tigerbeetle/pull/2221/commits/f6fa8466f242f544725e4f8941de44811f1c91e1